subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DependenciesExtensionModule.java

```diff
@@ -239,22 +242,83 @@ public static <D extends Dependency> void call(DependencyCollector self, D depen
         self.add(dependency, ConfigureUtil.configureUsing(configuration));
     }
 
+    @VisibleForTesting
+    public static final String ERROR_MESSAGE_PROVIDER = "Providers of non-Dependency(Constraint) types (java.lang.String) are not supported. Create a Dependency(Constraint) using Dependency(Constraint)Factory first.";
+
     /**
-     * Add a dependency.
+     * Add a dependency or dependency constraint.
      *
-     * @param dependency dependency to add
+     * <p>
+     * Unfortunately, this method cannot be split into two methods, one for {@link Dependency} and one for {@link DependencyConstraint},
+     * because Groovy cannot distinguish between the two.
+     * </p>
+     *
+     * @param dependencyOrDependencyConstraint dependency or dependency constraint to add
      */
-    public static void call(DependencyCollector self, Provider<? extends Dependency> dependency) {
-        self.add(dependency);
+    public static void call(DependencyCollector self, Provider<?> dependencyOrDependencyConstraint) {
+        self.add(dependencyOrDependencyConstraint.map(it -> {
+            if (it instanceof Dependency) {
+                return (Dependency) it;
+            }
+            if (it instanceof DependencyConstraint) {
+                return null;
+            }
+            throw new IllegalArgumentException(ERROR_MESSAGE_PROVIDER);
+        }));
+        self.addConstraint(dependencyOrDependencyConstraint.map(it -> {
+            if (it instanceof DependencyConstraint) {
+                return (DependencyConstraint) it;
+            }
+            if (it instanceof Dependency) {
+                return null;
+            }
+            throw new IllegalArgumentException(ERROR_MESSAGE_PROVIDER);
+        }));
     }
 
     /**
-     * Add a dependency.
+     * Add a dependency or dependency constraint.
      *
-     * @param dependency dependency to add
-     * @param configuration an action to configure the dependency
+     * @param dependencyOrDependencyConstraint dependency or dependency constraint to add
+     * @param configuration an action to configure the dependency or dependency constraint
      */
-    public static <D extends Dependency> void call(DependencyCollector self, Provider<? extends D> dependency, @ClosureParams(SecondParam.FirstGenericType.class) Closure<?> configuration) {
-        self.add(dependency, ConfigureUtil.configureUsing(configuration));
+    public static <D> void call(DependencyCollector self, Provider<? extends D> dependencyOrDependencyConstraint, @ClosureParams(SecondParam.FirstGenericType.class) Closure<?> configuration) {
```

Comment:
Is there any straightforward way to avoid duplicate logic between this method and the one above?

Reply:
I did this using `Closure.IDENTITY`.